### PR TITLE
[braft] New port

### PR DESCRIPTION
**Pull request by Gab-Menezes**
**Describe the pull request**
New port for [braft](https://github.com/baidu/braft)

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  Linux, No

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Not yet, I am still working on this PR

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Not yet, I am still working on this PR

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

**Comment by JackBoosY** [comment]
Can you please sign CLA first?

**Comment by Gab-Menezes** [comment]
Done

**Comment by Gab-Menezes** [comment]
If you remove the install step of the patch, the library will compile fine as I said. But we can't reference it in another cmake scripts

File: ports/braft/export-target.patch [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ else()
+-install(TARGETS braft-shared
++install(TARGETS braft-shared EXPORT unofficial-braftConfig
+         RUNTIME DESTINATION bin
+         LIBRARY DESTINATION lib${LIBSUFFIX}
+         ARCHIVE DESTINATION lib${LIBSUFFIX}
+         )
+ endif()
++
++install(EXPORT unofficial-braftConfig
```
Comment (JackBoosY) [inline ports/braft/export-target.patch:35]: All the dependency's styles are absolute path, so we don't need `find_dependency`.

**Comment by JackBoosY** [comment]
@Gab-Menezes Can you please test the changes?

**Comment by Gab-Menezes** [comment]
Hey, so with the changes it the install process went fine fine. So tried linking braft in test binary using
```cmake
cmake_minimum_required(VERSION 3.20)
project(test)

find_package(unofficial-braft CONFIG REQUIRED)

add_executable(main main.cpp)
target_link_libraries(main PRIVATE unofficial::braft::braft-static)
```
And the c++ program just have the main function returning 0.
But I got this error when trying to to compile (the configuration step went fine)
```
[ 50%] Building CXX object CMakeFiles/main.dir/main.cpp.o
make[2]: *** No rule to make target '/mnt/c/dev/vcpkg-dev/vcpkg/installed/x64-linux/debug/lib/libgflags.a', needed by 'main'. 
 Stop.
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/main.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```
If I try to link gflags manually
```cmake
cmake_minimum_required(VERSION 3.20)
project(test)

find_package(unofficial-braft CONFIG REQUIRED)
find_package(gflags CONFIG REQUIRED)

add_executable(main main.cpp)
target_link_libraries(main PRIVATE gflags::gflags)
target_link_libraries(main PRIVATE unofficial::braft::braft-static)
```
I get the same error, but if I remove the braft link and link only gflags everything is fine
```cmake
cmake_minimum_required(VERSION 3.20)
project(test)

find_package(gflags CONFIG REQUIRED)

add_executable(main main.cpp)
target_link_libraries(main PRIVATE gflags::gflags)
```
So I looked at the `lib` and `/debug/lib` folders
image
image
As we can see the `debug/lib` folder has a `_debug` suffix, but he expected name doesn't have this suffix.

After this can you talk me through what you have done, I would love to learn more and maybe create a PR to export brpc.

**Comment by JackBoosY** [comment]
> As we can see the debug/lib folder has a _debug suffix, but he expected name doesn't have this suffix.

We should find all kind of gflags.

> And got an error saying braft/raft.h is not in the include directories

Okay, will export only path level `include`.

**Comment by JackBoosY** [comment]
I will handle this tomorrow.

**Comment by JackBoosY** [comment]
@Gab-Menezes Please test again.

Thanks.

**Comment by Gab-Menezes** [comment]
So the configuration step is not working here is the log (running with --debug flag) (I alse deleted all the reference to braft to guarantee that it's not a cahe problem)
[config-x64-linux-dbg-err.log](https://github.com/microsoft/vcpkg/files/8403330/config-x64-linux-dbg-err.log)
[config-x64-linux-dbg-out.log](https://github.com/microsoft/vcpkg/files/8403331/config-x64-linux-dbg-out.log)
[console.txt](https://github.com/microsoft/vcpkg/files/8403332/console.txt)

**Comment by dg0yt** [comment]
~~~
grep: GFLAGS_INCLUDE_PATH-NOTFOUND/gflags/gflags_declare.h: No such file or directory
CMake Error at CMakeLists.txt:191 (if):
  if given arguments:

    "STREQUAL" "GFLAGS_NAMESPACE"

  Unknown arguments specified

~~~
This usually comes from a statement like:
~~~
if(${empyt_var} STREQUAL "GFLAGS_NAMESPACE")
~~~
Either you find out why it is empty/unset, or you add quotes:
~~~
if("${empyt_var}" STREQUAL "GFLAGS_NAMESPACE")
~~~

**Comment by JackBoosY** [comment]
@Gab-Menezes Should be good now.

**Comment by Gab-Menezes** [comment]
@JackBoosY Yay it compiles when linking to another project, only one thing. We need do link gflags manually if not we get:
```
CMake Error at /mnt/c/dev/vcpkg-dev/vcpkg/scripts/buildsystems/vcpkg.cmake:559 (_add_executable):
  Target "main" links to target "gflags::gflags" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  CMakeLists.txt:6 (add_executable)
```
Even though I didn't include nothing from glfags, but if we find and link, it works fine.

**Comment by Gab-Menezes** [comment]
In the near future I will try to do the brpc one (I'm kinda busy right now). If I have any problem can I ask for your help ?

**Comment by JackBoosY** [comment]
@Gab-Menezes Ah I forgot this, sorry.
We already have port grpc. can you please check whether it suitable for you?

**Comment by Gab-Menezes** [comment]
Yes I know we have, but it's not exported

**Comment by JackBoosY** [comment]
> Yes I know we have, but it's not exported (I'm talking about brpc (braft uses brpc), not grpc)

Okay, I will take a look at that.

**Comment by JackBoosY** [comment]
@Gab-Menezes This time it should be okay, I promised.

**Comment by Gab-Menezes** [comment]
@JackBoosY It's working perfectly without needing to link gflags manually.

**Review by JackBoosY — APPROVED** [review]

**Comment by strega-nil-ms** [comment]
Thanks!

**Pull request merged**